keps/sig-scheduling/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,885 @@
+    // feature gate.
+    //
+    // +optional
+    // +featureGate:DRAAdminControlledDeviceAttributes
+    Capacity map[FullyQualifiedName]DeviceCapacity
+
+    // ^^^^
+    // The assumption here is that all device types will have attributes and capacities,
+    // similar to the current BasicDevice type. Therefore the overrides are not made
+    // specific to certain device types.
+}
+
+// DevicePatchFilter defines which device(s) a [DevicePatch] applies to.
+type DevicePatchFilter struct {
+    // If DeviceClassName is set, the selectors defined there must be
+    // satisfied by a device to be patched. This field corresponds
+    // to class.metadata.name.
+    //
+    // +optional
+    DeviceClassName *string
+
+    // If driver is set, only devices from that driver are patched.
+    // This fields corresponds to slice.spec.driver.
+    //
+    // +optional
+    Driver *string
+
+    // If pool is set, only devices in that pool are patched.
+    // This fields corresponds to slice.spec.pool.name.
+    //
+    // Also setting the driver name may be useful to avoid
+    // ambiguity when different drivers use the same pool name,
+    // but this is not required because selecting pools from
+    // different drivers may also be useful, for example when
+    // drivers with node-local devices use the node name as
+    // their pool name.
+    //
+    // +optional
+    Pool *string
+
+    // If device is set, only devices with that name are patched.
+    // This field corresponds to slice.spec.devices[].name.
+    //
+    // Also setting driver and pool may be required to avoid ambiguity,
+    // but is not required.
+    //
+    // +optional
+    Device *string
+
+    // Selectors define criteria which must be satisfied by a
+    // device to be patched. All selectors must be satisfied.
+    //
+    // +optional
+    // +listType=atomic
+    Selectors []DeviceSelector
+
+    // ^^^
+    //
+    // Selectors is a list for the same reason why a request has a list: at some point
+    // we might have one entry which uses CEL while another uses a different, yet to be
+    // defined mechanism.
+}
+```
+
+To distinguish intentionally empty attributes from attributes which have some
+future, unknown content and thus only seem to be empty to an older client, a
+special "null value" gets introduced:
+
+```golang
+// NullableDeviceAttribute must have exactly one field set.
+// It has the exact same fields as a [DeviceAttribute] plus NullValue as
+// an additional alternative.
+type NullableDeviceAttribute struct {
+    DeviceAttribute
+
+    // NullValue, if set, marks an intentionally empty attribute.
+    //
+    // +optional
+    // +oneOf=ValueType
+    NullValue *NullValue `json:"null,omitempty"`
+}
+
+// ^^^
+// `NullableDeviceAttribute` as an extension ensures that the OpenAPI
+// for ResourceSlice remains unchanged. Using the same type with
+// a `NullValue` that can be set only in one type is less clear.
```

Comment:
@aaron-prindle: will a NullableDeviceAttribute with some "oneOf" alternatives inside the embedded DeviceAttribute and one more outside of it work for [declarative validation](https://github.com/kubernetes/enhancements/pull/5074/files#diff-ee5e894a755c55c5e78fa01f4d06cfc09d7ea1868f3c6ee1e905da474fe00352)?

It should work right now (OpenAPI flattens embedded structs) and it is more natural in Go (can use a NullableDeviceAttribute to initialize a DeviceAttribute without manually written copy code).

But if this then poses a problem for declarative validation, then it will be difficult to switch because the embedding leads to different protobuf encoding.

Reply:
Yes, exactly, hence the explicit "null". One alternative in a prior comment thread in this PR was an explicit `remove: true` (but then what does `remove: false` mean?) or `remove: {}`.